### ISSUE TITLE
Not working

### ISSUE THREAD

**Issue by yesnoyes** [aleemrazzaq/CSR_EC_256#1]
Hi Razzaq,
I am trying to generate CSR with ECC 256 bits. I got the following output which is incorrect. Is it working for you?

-----BEGIN CERTIFICATE REQUEST-----
MIHHMG8CAQAwJzEOMAwGA1UEAwwFUmF6ZXIxFTATBgNVBAUMDDQ4NDkyOTQ1ODc1MAQJ2tkvC7d4gYW2uq7cIfRGKXcGf6/Lw+NK/QVVvNs9lGgQ+PUmwIIXt2Nm6BCtfaHe+YfvZJQl/dH5xrjKdinkoAAwCgYIKoZIzj0EAwIDSAAwRQIhAMkkIPF9ph9/RC2yeWn3RNQCq/SVcwGYWyWe60AS1Eq3AiBV1HDjxgTAr++ciF8edouyqU+gf5tOevS61Axs8RKF7w==
-----END CERTIFICATE REQUEST-----

**Comment by aleemrazzaq** [comment]
yes i looked at your output and its not accurate, what are you passing `dataToBeSigned` ? And `privateKey` fetching from secure enclave or just in keychain ?

**Comment by yesnoyes** [comment]
Hi Aleem,
Thanks for quick response. Here is the swift code i am using to generate key pair - 

    public static func generateKeyPair(ofSize bits:UInt) throws -> KeyPair {
        let pubKeyAttrs = [ kSecAttrIsPermanent as String: true ]
        let privKeyAttrs = [ kSecAttrIsPermanent as String: true ]
        let params: NSDictionary = [ kSecAttrKeyType as String : kSecAttrKeyTypeEC as String,
                       kSecAttrKeySizeInBits as String : bits, // I am passing 256
                       kSecPublicKeyAttrs as String : pubKeyAttrs,
                       kSecPrivateKeyAttrs as String : privKeyAttrs ]
        var pubKey: SecKey?
        var privKey: SecKey?
        let status = SecKeyGeneratePair(params, &pubKey, &privKey)
        
        guard status == errSecSuccess else {
            throw SecKeyError.generateKeyPairFailed(osStatus: status)
        }
        guard let pub = pubKey, let priv = privKey else {
            throw SecKeyError.generateKeyPairFailed(osStatus: nil)
        }

        try changeKeyTag(priv)
        try changeKeyTag(pub)

        return (priv, pub)
    }

**Comment by yesnoyes** [comment]
Do you have an example program which is working with ECC?

**Comment by yesnoyes** [comment]
I am passing "public key bits" for data to be signed

**Comment by aleemrazzaq** [comment]
Thanks for sharing code. Yes this is for ECC and tested too.
If you're passing public key then that should be in DER encoded, better to follow or get help from this [https://github.com/DigitalLeaves/CryptoExportImportManager](url). This man is creating keypair in Secure Enclave and then convert public key in DER encoded. Kindly do let me know about progress after try?

**Comment by aleemrazzaq** [comment]
Kindly get updated code for working private key which store in Secure Enclave.

**Comment by yesnoyes** [comment]
Hi Aleem,
I tried everything as you said and also used https://github.com/DigitalLeaves/CryptoExportImportManager for generating keypair and converting public key to DER to pass into data but same problem. Do you have a sample code which can generate ECC CSR?
Appreciate and Thanks!

**Comment by cbaker6** [comment]
@yesnoyes I ran into some of the same issues you had with this code. The implementation to get everything working isn't is straight forward as it seems.

I have implemented a Swift version which works for ECC and RSA here: https://github.com/cbaker6/CertificateSigningRequestSwift

**Comment by aleemrazzaq** [comment]
@yesnoyes Sorry for late, I was quite busy from last few months, checkout this repo https://github.com/aleemrazzaq/iOSCSR
This is for both EC and RSA plus how to use in code, the description is also there.
